File: src/main/java/com/company/erp/features/users/service/RoleService.java
package com.company.erp.features.users.service;

import com.company.erp.core.exceptions.ConflictException;
import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.features.users.dto.CreateRoleRequest;
import com.company.erp.features.users.dto.UpdateRoleRequest;
import com.company.erp.features.users.entity.Permission;
import com.company.erp.features.users.entity.Role;
import com.company.erp.features.users.repository.PermissionRepository;
import com.company.erp.features.users.repository.RoleRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@Transactional
public class RoleService {

    private final RoleRepository roles;
    private final PermissionRepository permissions;

    public RoleService(RoleRepository roles, PermissionRepository permissions) {
        this.roles = roles;
        this.permissions = permissions;
    }

    @Transactional(readOnly = true)
    public List<Role> list() {
        return roles.findAll();
    }

    @Transactional(readOnly = true)
    public Role getById(Long id) {
        return roles.findById(id).orElseThrow(() -> new NotFoundException("Role not found"));
    }

    @Transactional(readOnly = true)
    public List<Permission> listPermissions() {
        return permissions.findAll();
    }

    public Role create(CreateRoleRequest req) {
        if (roles.existsByCode(req.code())) throw new ConflictException("Role code already in use");
        Role role = new Role();
        role.setCode(req.code());
        role.setName(req.name());
        role.setDescription(req.description());
        role.setPermissions(new HashSet<>(resolvePermissions(req.permissions())));
        return roles.save(role);
    }

    public Role update(Long id, UpdateRoleRequest req) {
        Role role = getById(id);
        if (req.name()        != null) role.setName(req.name());
        if (req.description() != null) role.setDescription(req.description());
        if (req.permissions() != null) role.setPermissions(new HashSet<>(resolvePermissions(req.permissions())));
        return role;
    }

    public void delete(Long id) {
        roles.delete(getById(id));
    }

    private Set<Permission> resolvePermissions(Set<String> codes) {
        if (codes.isEmpty()) return Set.of();
        List<Permission> found = permissions.findByCodeIn(codes);
        Set<String> foundCodes = found.stream().map(Permission::getCode).collect(Collectors.toSet());
        Set<String> missing = new HashSet<>(codes);
        missing.removeAll(foundCodes);
        if (!missing.isEmpty()) {
            throw new NotFoundException("Unknown permission(s): " + String.join(",", missing));
        }
        return new HashSet<>(found);
    }
}

File: app/Features/Users/Services/RoleService.php
<?php

namespace App\Features\Users\Services;

use App\Features\Users\Models\Permission;
use App\Features\Users\Models\Role;
use App\Support\Exceptions\ConflictException;
use App\Support\Exceptions\NotFoundException;
use Illuminate\Support\Facades\DB;

class RoleService
{
    /**
     * @return \Illuminate\Database\Eloquent\Collection<int, Role>
     */
    public function list()
    {
        return Role::query()->with('permissions')->get();
    }

    public function getById(int $id): Role
    {
        $role = Role::query()->with('permissions')->find($id);
        if (! $role) {
            throw new NotFoundException('Role not found');
        }
        return $role;
    }

    /**
     * @return \Illuminate\Database\Eloquent\Collection<int, Permission>
     */
    public function listPermissions()
    {
        return Permission::query()->get();
    }

    /**
     * @param array<string, mixed> $data  code,name,description,permissions
     */
    public function create(array $data): Role
    {
        if (Role::query()->where('code', $data['code'])->exists()) {
            throw new ConflictException('Role code already in use');
        }

        return DB::transaction(function () use ($data) {
            $role = new Role();
            $role->code = $data['code'];
            $role->name = $data['name'];
            $role->description = $data['description'] ?? null;
            $role->save();

            $permissions = $this->resolvePermissions($data['permissions'] ?? []);
            $role->permissions()->sync($permissions->pluck('id')->all());

            return $role->load('permissions');
        });
    }

    /**
     * @param array<string, mixed> $data  partial: name,description,permissions
     */
    public function update(int $id, array $data): Role
    {
        return DB::transaction(function () use ($id, $data) {
            $role = $this->getById($id);

            if (array_key_exists('name', $data) && $data['name'] !== null) {
                $role->name = $data['name'];
            }
            if (array_key_exists('description', $data) && $data['description'] !== null) {
                $role->description = $data['description'];
            }
            $role->save();

            if (array_key_exists('permissions', $data) && $data['permissions'] !== null) {
                $permissions = $this->resolvePermissions($data['permissions']);
                $role->permissions()->sync($permissions->pluck('id')->all());
            }

            return $role->load('permissions');
        });
    }

    public function delete(int $id): void
    {
        $role = $this->getById($id);
        $role->delete();
    }

    /**
     * @param array<int, string> $codes
     * @return \Illuminate\Support\Collection<int, Permission>
     */
    private function resolvePermissions(array $codes)
    {
        if (empty($codes)) {
            return collect();
        }
        $found = Permission::query()->whereIn('code', $codes)->get();
        $foundCodes = $found->pluck('code')->all();
        $missing = array_values(array_diff(array_values(array_unique($codes)), $foundCodes));
        if (! empty($missing)) {
            throw new NotFoundException('Unknown permission(s): '.implode(',', $missing));
        }
        return $found;
    }
}
